### ISSUE TITLE
Term::Complete screws up -ixon

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4460]
Migrated from [rt.perl.org#7765](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7765) (status was 'resolved')

Searchable as RT7765$

**Comment by p5pRT** [comment]
### From @nwc10



Aaaaaaaaaaaaaaagh\.
I've been trying to work out what's messing up my terminal\. It turns out
that it's the addition of a test for Term&#8203;::Complete\.
Term&#8203;::Complete&#8203;::Compete doesn't perfectly restore my terminal\. It turns
\-ixon to ixon\.
^S and ^Q become flow control\. \*not\* what I started with

Raw xterm grab\, which superficially looks like a lot of garbage&#8203;:

nick@&#8203;Bagpuss \[t\]$ stty \-all
speed 9600 baud; rows 50; columns 80; line = 0;
intr = ^C; quit = ^\\; erase = ^?; kill = ^U; eof = ^D; eol = \<undef>;
eol2 = \<undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R; werase = ^W;
lnext = ^V; flush = ^O; min = 1; time = 0;
\-parenb \-parodd cs8 \-hupcl \-cstopb cread \-clocal \-crtscts
\-ignbrk brkint ignpar \-parmrk \-inpck istrip \-inlcr \-igncr icrnl \-ixon \-ixoff
\-iuclc \-ixany \-imaxbel
opost \-olcuc \-ocrnl onlcr \-onocr \-onlret \-ofill \-ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok \-echonl \-noflsh \-xcase \-tostop \-echoprt
echoctl echoke
nick@&#8203;Bagpuss \[t\]$ perl5\.6\.1 \-MTerm&#8203;::Complete \-e 'Complete "perl "\, "rules"'
perl rules
nick@&#8203;Bagpuss \[t\]$ stty \-all
speed 9600 baud; rows 50; columns 80; line = 0;
intr = ^C; quit = ^\\; erase = ^?; kill = ^U; eof = ^D; eol = \<undef>;
eol2 = \<undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R; werase = ^W;
lnext = ^V; flush = ^O; min = 1; time = 0;
\-parenb \-parodd cs8 \-hupcl \-cstopb cread \-clocal \-crtscts
\-ignbrk brkint ignpar \-parmrk \-inpck istrip \-inlcr \-igncr icrnl ixon \-ixoff
\-iuclc \-ixany \-imaxbel
opost \-olcuc \-ocrnl onlcr \-onocr \-onlret \-ofill \-ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok \-echonl \-noflsh \-xcase \-tostop \-echoprt
echoctl echoke

First stty \-all shows the state of my terminal
Then I run perl5\.6\.1 \-MTerm&#8203;::Complete \-e 'Complete "perl "\, "rules"'
Second stty \-all shows the new state of my terminal

or more obviously&#8203;:

nick@&#8203;Bagpuss \[t\]$ stty \-all | grep ixon | sed \-e 's/\.\*\\\(\.ixon\\\)\.\*/\\1/'
\-ixon
nick@&#8203;Bagpuss \[t\]$ perl5\.6\.1 \-MTerm&#8203;::Complete \-e 'Complete "perl "\, "rules"'
perl rules
nick@&#8203;Bagpuss \[t\]$ stty \-all | grep ixon | sed \-e 's/\.\*\\\(\.ixon\\\)\.\*/\\1/'
 ixon

only change is that \-ixon has become ixon\.
Baaah\.

I'm not enough of a terminal wizard to know why the routines&#8203;:
&nbsp;        $tty\_raw\_noecho = "$stty raw \-echo";
&nbsp;        $tty\_restore    = "$stty \-raw echo";

don't do what they should\, or whether a stty \-all can be used instead to
capture terminal output and restore it perfectly \(so sorry\, no patch\)\, but
I'd hope that something can be done to fix this irritating bug\.
I like my forward search in bash\, and I don't like no steenkin' test turning
it into "mysteriously\-lock\-up\-my\-terminal"
\(you may set your terminal how you like\. This is none of my business\)

bug in 5\.6\.1 and more annoyingly in bleadperl\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=lib
    severity=low

This perlbug was built using Perl v5.7.2 - Sat Sep 15 14:10:21 BST 2001
It is being executed now by  Perl v5.7.2 - Sat Sep 15 17:02:46 BST 2001.

Site configuration information for perl v5.7.2:

Configured by nick at Sat Sep 15 17:02:46 BST 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 2 patch 12025) configuration:
  Platform:
    osname=linux, osvers=2.4.5-rmk6, archname=armv4l-linux
    uname='linux bagpuss.unfortu.net 2.4.5-rmk6 #4 tue jun 19 23:26:57 bst 2001 armv4l unknown '
    config_args='-Doptimize=-O1 -Dusedevel -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@talking.bollo.cx -Dperladmin=nick@talking.bollo.cx -Dinc_version_list=  -Dinc_version_list_init=0 -Uuse64bitint -de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O1',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20010319 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL12022


@INC for perl v5.7.2:
    /usr/local/lib/perl5/5.7.2/armv4l-linux
    /usr/local/lib/perl5/5.7.2
    /usr/local/lib/perl5/site_perl/5.7.2/armv4l-linux
    /usr/local/lib/perl5/site_perl/5.7.2
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.2:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_GB.ISO-8859-1
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On Solaris and Linux\, you can use \`stty \-g\`\. This outputs a string that can
be given back as parameter to stty to restore the terminal state\.
But I don't know whether it's portable\.

**Comment by p5pRT** [comment]
### From @jhi

I guess one could try running stty \-g\, if succesful \(no error code\)\,
save away the output\, and on completion restore using that\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> I guess one could try running stty \-g\, if succesful \(no error code\)\, save away
> the output\, and on completion restore using that\.

Something like this?

\-\- c


<details><summary>Inline Patch</summary>

```diff
--- lib/Term/~Complete.t	Sun Sep 30 18:13:15 2001
+++ lib/Term/Complete.t	Sun Sep 30 18:13:17 2001
@@ -8,6 +8,11 @@
 use warnings;
 use Test::More tests => 8;
 use vars qw( $Term::Complete::complete $complete );
+my $restore;
+
+END {
+	`stty $restore`;
+}
 
 SKIP: {
     skip('PERL_SKIP_TTY_TEST', 8) if $ENV{PERL_SKIP_TTY_TEST};
@@ -18,7 +23,9 @@
     if (defined $TTY) {
 	open(TTY, $TTY)               or die "open $TTY failed: $!";
 	skip("$TTY not a tty", 8)     if defined $TTY && ! -t TTY;
-    }
+	$restore = `stty -g`;
+	skip("Can't reliably restore $TTY", 8) if $?;
+	}
 
 use_ok( 'Term::Complete' );
```
</details>

**Comment by p5pRT** [comment]
### From @doughera88

> On 2001\.09\.30 01&#8203;:07 Nicholas Clark wrote&#8203;:
> > 
> > I'm not enough of a terminal wizard to know why the routines&#8203;:
> >         $tty\_raw\_noecho = "$stty raw \-echo";
> >         $tty\_restore    = "$stty \-raw echo";
> > 
> > don't do what they should\, or whether a stty \-all can be used instead to
> > capture terminal output and restore it perfectly \(so sorry\, no patch\)\, but
> > I'd hope that something can be done to fix this irritating bug\.
 
Here's the relevant section from the OpenBSD 2\.9 stty\(1\) man page&#8203;:

raw \(\-raw\)  If set\, change the modes of the terminal so that no input or
&nbsp;            output processing is performed\.  If unset\, change the modes
&nbsp;            of the terminal to some reasonable state that performs input
&nbsp;            and output processing\.  Note that since the terminal driver
&nbsp;            no longer has a single RAW bit\, it is not possible to intuit
&nbsp;            what flags were set prior to setting raw\. This means that un\-
&nbsp;            setting raw may not put back all the setting that were previ\-
&nbsp;            ously in effect\.  To set the terminal into a raw state and
&nbsp;            then accurately restore it\, the following shell code is rec\-
&nbsp;            ommended&#8203;:

&nbsp;                  save\_state=$\(stty \-g\)
&nbsp;                  stty raw
&nbsp;                  \.\.\.
&nbsp;                  stty "$save\_state"

> On Solaris and Linux\, you can use \`stty \-g\`\. This outputs a string that can
> be given back as parameter to stty to restore the terminal state\.
> But I don't know whether it's portable\.

Alas\, stty \-g is not portable\.  Stty stuff is notoriously non\-portable\.
BSD and SysV have traditionally been nearly completely different universes
when it comes to stty\. Stty \-g comes from the SysV end of the universe\.  
Newer BSD\-derived systems \(such as all the various free \*BSD systems\)
probably have it\, but older ones may well not\. A further complication
\(perhaps even with the above code\) is that traditional BSD stty operated
on STDOUT\, while SysV stty operates on STDIN\.

I'd recommend only even attempting the Term&#8203;::Complete test if stty \-g is
successful\. If it isn't successful\, you can't be sure you can reset the
terminal state and hence you shouldn't mess it up\.  Lastly\, note that
older BSD\-derived systems have traditionally been under\-represented on
p5p\, so tread carefully\.

**Comment by p5pRT** [comment]
### From @jhi

That's the latest iteration that chromatic sent to me\.

> terminal state and hence you shouldn't mess it up\.  Lastly\, note that
> older BSD\-derived systems have traditionally been under\-represented on
> p5p\, so tread carefully\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Nothing using stty seems to be highly portable\.

It seems to me that instead of having code scatered throughout Term&#8203;::
and elsewhere which calls on the stty program\, we should have a pragma
or module which encapsulates the functionality\.  This one module could
be configured appropriately on a per\-system basis\, so as to try really
hard \[at installation time\, not at runtime\]\, to get it right\.

use IO&#8203;::Stty;
\{ my $save = stty\(qw\(raw \-echo\)\);
&nbsp;  \# do stuff with raw on and echo off\.
&nbsp;  $save\->restore;
\}

Or something like that\.\.\.

Besides looking at stty and seeing what arguments it accepts\, it could
also try and use ioctl\, which is even more nonportable\, but which should
be much faster than calling an external program\.

**Comment by p5pRT** [comment]
### From @samtregar

> It seems to me that instead of having code scatered throughout Term&#8203;::
> and elsewhere which calls on the stty program\, we should have a pragma
> or module which encapsulates the functionality\.  This one module could
> be configured appropriately on a per\-system basis\, so as to try really
> hard \[at installation time\, not at runtime\]\, to get it right\.

You mean something like the existing IO&#8203;::Stty perhaps?

&nbsp;   http&#8203;://search\.cpan\.org/search?dist=IO\-Stty

\-sam

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hmm\, yes precisely\.

Though looking at how the program does what it does\, I'd probably write
much of it differently \.\.\. like using hashes rather than seperate
variables for $c\_cflag\, $c\_iflag\, $ispeed\, etc\, using for loops instead
of ugly blocks like&#8203;:
&nbsp;  $control\_chars\{'INTR'\}=$termios\->getcc\(VINTR\);
&nbsp;  $control\_chars\{'QUIT'\}=$termios\->getcc\(VQUIT\);
&nbsp;  $control\_chars\{'ERASE'\}=$termios\->getcc\(VERASE\);
&nbsp;  $control\_chars\{'KILL'\}=$termios\->getcc\(VKILL\);
&nbsp;  $control\_chars\{'EOF'\}=$termios\->getcc\(VEOF\);
&nbsp;  $control\_chars\{'TIME'\}=$termios\->getcc\(VTIME\);
&nbsp;  $control\_chars\{'MIN'\}=$termios\->getcc\(VMIN\);
&nbsp;  $control\_chars\{'START'\}=$termios\->getcc\(VSTART\);
&nbsp;  $control\_chars\{'STOP'\}=$termios\->getcc\(VSTOP\);
&nbsp;  $control\_chars\{'SUSP'\}=$termios\->getcc\(VSUSP\);
&nbsp;  $control\_chars\{'EOL'\}=$termios\->getcc\(VEOL\);
and other similar blocks\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Sep 29 09&#8203;:07&#8203;:19 2001\, nicholas wrote&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Aaaaaaaaaaaaaaagh\.
> I've been trying to work out what's messing up my terminal\. It turns
>    out
> that it's the addition of a test for Term&#8203;::Complete\.
> Term&#8203;::Complete&#8203;::Compete doesn't perfectly restore my terminal\. It
>    turns
> \-ixon to ixon\.
> ^S and ^Q become flow control\. \*not\* what I started with
> 

Are these issues still present with the version of Term&#8203;::Complete
currently included with Perl?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Mar 28\, 2012 at 07&#8203;:33&#8203;:19PM \-0700\, James E Keenan via RT wrote&#8203;:
> On Sat Sep 29 09&#8203;:07&#8203;:19 2001\, nicholas wrote&#8203;:
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > Aaaaaaaaaaaaaaagh\.
> > I've been trying to work out what's messing up my terminal\. It turns
> >    out
> > that it's the addition of a test for Term&#8203;::Complete\.
> > Term&#8203;::Complete&#8203;::Compete doesn't perfectly restore my terminal\. It
> >    turns
> > \-ixon to ixon\.
> > ^S and ^Q become flow control\. \*not\* what I started with
> > 
> 
> Are these issues still present with the version of Term&#8203;::Complete
> currently included with Perl?

Good question\.

Gosh\. It turns out that this bug is long fixed\. The fix was&#8203;:

commit fc81c880d856f71fc84201fd183d90bfd812c5c2
Author&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi>
Date&#8203;:   Mon Oct 1 01&#8203;:19&#8203;:21 2001 \+0000

&nbsp;    Use stty \-g to restore the terminal state\, do not even
&nbsp;    try to run the tests if stty \-g is not available\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;12282

which was approximately 2 days after it was reported\.
\(as ID 20010929\.014 in the old system
https://rt.perl.org/cgi-bin/pb2rt.cgi?PBID=20010929.014 \)

Thanks for digging this one up\. I'm about to mark it as resolved\.

Nicholas Clark
